lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
-
-            _serviceAndContextConfigurationProvider.EvaluatorFailed(value.Id);
+            _serviceAndContextConfigurationProvider.RecordEvaluatorFailureById(value.Id);
+            bool isMaster = _serviceAndContextConfigurationProvider.IsMasterEvaluatorId(value.Id);
 
-            // if active context stage is reached for Update Task then assume that failed
-            // evaluator belongs to mapper
-            if (_reachedUpdateTaskActiveContext)
+            // If failed evaluator is master then ask for master 
+            // evaluator else ask for mapper evaluator
+            if (!isMaster)
             {
+                Logger.Log(Level.Info, "Requesting for the failed map evaluator again");
                 RequestMapEvaluators(1);
             }
             else
             {
+                Logger.Log(Level.Info, "Requesting for the failed master evaluator again");
                 RequestUpdateEvaluator();
             }
         }
 
         /// <summary>
-        /// Specfies how to handle exception or error
+        /// Specifies what to do if Failed Context is received.
+        /// An exception is thrown if tasks are not completed.
         /// </summary>
-        /// <param name="error">Kind of exception</param>
-        public void OnError(Exception error)
+        /// <param name="value"></param>
+        public void OnNext(IFailedContext value)
+        {
+            if (_areImruTasksCompleted)
+            {
+                Logger.Log(Level.Info,
+                    string.Format("Context with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
+                return;
+            }
+            Exceptions.Throw(new Exception(string.Format("Data Loading Context with Id: {0} failed", value.Id)), Logger);
+        }
+
+        /// <summary>
+        /// Specifies what to do if a task fails.
+        /// We throw the exception and fail IMRU unless IMRU job is already done.
+        /// </summary>
+        /// <param name="value"></param>
+        public void OnNext(IFailedTask value)
         {
-            Logger.Log(Level.Error, "Cannot currently handle the Exception in OnError function");
-            throw new NotImplementedException("Cannot currently handle exception in OneError", error);
+            if (_areImruTasksCompleted)
+            {
+                Logger.Log(Level.Info,
+                    string.Format("Task with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
+                return;
+            }
+            Exceptions.Throw(new Exception(string.Format("Task with Id: {0} failed", value.Id)), Logger);
         }
 
         /// <summary>
-        /// Specfies what to do when driver is done
-        /// In this case do nothing
+        /// Specifies what to do once we receive a running task
         /// </summary>
+        /// <param name="value"></param>
+        public void OnNext(IRunningTask value)
```

Comment:
This is the same behavior as if IMRU wouldn't bind a handler for `IRunningTask` at all. Probably safe to remove altogether.